docs/cloud-integration.md

```diff
@@ -0,0 +1,512 @@
+1. Have the Spark context configured with the authentication details of the object store.
+In a YARN cluster, this may also be done in the `core-site.xml` file.
+
+
+## <a name="output"></a>Object Stores as a substitute for HDFS
+
+As the examples show, you can write data to object stores. However, that does not mean
+That they can be used as replacements for a cluster-wide filesystem.
+
+The full details are covered in [Cloud Object Stores are Not Real Filesystems](#cloud_stores_are_not_filesystems).
+
+The brief summary is:
+
+| Object Store Connector      |  Replace HDFS? |
+|-----------------------------|--------------------|
+| `s3a://` `s3n://`  from the ASF   | No  |
+| Amazon EMR `s3://`          | Yes |
+| Microsoft Azure `wasb://`   | Yes |
+| OpenStack `swift://`        | No  |
+
+It is possible to use any of the object stores as a destination of work, i.e. use
+`saveAsTextFile()` or `save()` to save data there, but the commit process may be slow
+and, unreliable in the presence of failures.
+
+It is faster and safer to use the cluster filesystem as the destination of Spark jobs,
+using that data as the data for follow-on work. The final results can
+be persisted in to the object store using `distcp`.
+
+#### <a name="checkpointing"></a>Spark Streaming and object stores
+
+Spark Streaming can monitor files added to object stores, by
+creating a `FileInputDStream` DStream monitoring a path under a bucket through
+`StreamingContext.textFileStream()`.
+
+
+1. The time to scan for new files is proportional to the number of files
+under the path —not the number of *new* files, and that it can become a slow operation.
+The size of the window needs to be set to handle this.
+
+1. Files only appear in an object store once they are completely written; there
+is no need for a worklow of write-then-rename to ensure that files aren't picked up
+while they are still being written. Applications can write straight to the monitored directory.
+
+1. Streams should only be checkpointed to an object store considered compatible with
+HDFS. Otherwise the checkpointing will be slow and potentially unreliable.
+
+### Recommended settings for writing to object stores
+
+Here are the settings to use when writing to object stores. This uses the "version 2" algorithm
+for committing files —which does less renaming than the v1 algorithm. Speculative execution is
+disabled to avoid multiple writers corrupting the output.
+
+```
+spark.speculation false
+spark.hadoop.mapreduce.fileoutputcommitter.algorithm.version 2
+spark.hadoop.mapreduce.fileoutputcommitter.cleanup-failures.ignored true
+```
+
+There's also the option of skipping the cleanup of temporary files in the output directory.
+Enabling this option eliminates a small delay caused by listing and deleting any such files.
+
+```
+spark.hadoop.mapreduce.fileoutputcommitter.cleanup.skipped true
+```
+
+Bear in mind that storing temporary files can run up charges; Delete
```

Comment:
Yeah, is this a good idea? sooner or later these need to be cleaned up, and can it be that expensive? I just question whether to call this "recommended"